Web/Scotty/Action.hs

```diff
@@ -226,19 +233,86 @@ jsonData = do
 --
--- * If parameter is found, but 'read' fails to parse to the correct type, 'next' is called.
+-- * If parameter is found, but 'parseParam' fails to parse to the correct type, 'next' is called.
 --   This means captures are somewhat typed, in that a route won't match if a correctly typed
 --   capture cannot be parsed.
 param :: (Parsable a, ScottyError e, Monad m) => T.Text -> ActionT e m a
-param k = do
-    val <- ActionT $ liftM (lookup k . getParams) ask
-    case val of
-        Nothing -> raise $ stringError $ "Param: " ++ T.unpack k ++ " not found!"
-        Just v  -> either (const next) return $ parseParam v
+param = paramWith Nothing getParams status500
+{-# DEPRECATED param "(#204) Not a good idea to treat all parameters identically. Use captureParam, formParam and queryParam instead. "#-}
+
+-- | Get a capture parameter.
+--
+-- * Raises an exception which can be caught by 'rescue' if parameter is not found. If the exception is not caught, scotty will return a HTTP error code 500 ("Internal Server Error") to the client.
+--
+-- * If parameter is found, but 'parseParam' fails to parse to the correct type, 'next' is called.
+captureParam :: (Parsable a, ScottyError e, Monad m) => T.Text -> ActionT e m a
+captureParam = paramWith (Just CaptureParam) getCaptureParams status500
+
+-- | Get a form parameter.
+--
+-- * Raises an exception which can be caught by 'rescue' if parameter is not found. If the exception is not caught, scotty will return a HTTP error code 400 ("Bad Request") to the client.
+--
+-- * If parameter is found, but 'parseParam' fails to parse to the correct type, 'next' is called.
+formParam :: (Parsable a, ScottyError e, Monad m) => T.Text -> ActionT e m a
+formParam = paramWith (Just FormParam) getFormParams status400
+
+-- | Get a query parameter.
+--
+-- * Raises an exception which can be caught by 'rescue' if parameter is not found. If the exception is not caught, scotty will return a HTTP error code 400 ("Bad Request") to the client.
+--
+-- * If parameter is found, but 'parseParam' fails to parse to the correct type, 'next' is called.
+queryParam :: (Parsable a, ScottyError e, Monad m) => T.Text -> ActionT e m a
+queryParam = paramWith (Just QueryParam) getQueryParams status400
+
+data ParamType = CaptureParam
+               | FormParam
+               | QueryParam
+instance Show ParamType where
+  show = \case
+    CaptureParam -> "capture"
+    FormParam -> "form"
+    QueryParam -> "query"
+
+paramWith :: (ScottyError e, Monad m, Parsable b) =>
+             Maybe ParamType
+          -> (ActionEnv -> [Param])
+          -> Status -- ^ HTTP status to return if parameter is not found
+          -> T.Text -- ^ parameter name
+          -> ActionT e m b
+paramWith tym f err k = do
+    val <- ActionT $ liftM (lookup k . f) ask
+    case tym of
+      Nothing -> raiseStatus err $ stringError (unwords ["Parameter:", T.unpack k, "not found!"])
```

Comment:
It looks like we always throw in the case where `tym` is `Nothing`, if I understood correctly this is what is called when the user calls the deprecated `param` we should try to preserve the old behavior which is essentially the same as what we now do for `captureParam` (error500 when not found and next where no parse)

Reply:
good point! I'll revert "param" to the original implementation until we get rid of it :)